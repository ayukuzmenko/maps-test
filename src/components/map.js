import React, { Component } from 'react';
import { connect } from 'react-redux';

class Map extends Component {
  state = {
    map: null,
  };
  initMap = () => {
    const { ymaps } = this.props;
    const map = new ymaps.Map(
      'map',
      {
        center: [55.76, 37.64], // Москва
        zoom: 10,
      },
      {
        searchControlProvider: 'yandex#search',
      },
    );

    this.setState({ map });
  };

  render() {
    const mapHeight = window.innerHeight - 50;
    return <div id="map" style={{ width: '70%', height: mapHeight }} />;
  }

  componentDidMount() {
    const { ymaps } = this.props;
    ymaps.ready(this.initMap);
  }
}

const mapStateProps = state => {
  return {
    ymaps: state.ymaps,
  };
};
export default connect(mapStateProps)(Map);
