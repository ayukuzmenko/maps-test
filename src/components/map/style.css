.containerMap {
  position: absolute;
  right: 0;
  width: 70%;
  height: 90%;
}
