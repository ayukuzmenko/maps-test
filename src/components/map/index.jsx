import React, { Component } from 'react';
import { connect } from 'react-redux';
import './style.css';

export class Map extends Component {
  initMap = () => {
    this._map = new window.ymaps.Map(
      'map',
      {
        center: [55.76, 37.64], // Москва
        zoom: 10,
      },
      {
        searchControlProvider: 'yandex#search',
      },
    );
  };

  handlerDragEnd = event => {
    const mark = event.get('target');
    const { replacePoint } = this.props;
    const newCoords = mark.geometry.getCoordinates();
    const arrIndex = mark.properties.get('arrIndex');
    replacePoint({ arrIndex, newCoords });
  };

  render() {
    return <div id="map" className="containerMap" />;
  }

  componentDidMount() {
    window.ymaps.ready(this.initMap);
  }

  componentDidUpdate() {
    this._map.geoObjects.removeAll();
    this.addPoint();
    this.addPolyLine();
  }

  addPoint() {
    const { points } = this.props;
    
    if (points.length > 0) {
      points.forEach((point, index) => {
        const mark = new window.ymaps.Placemark(
          point.coords,
          {
            iconCaption: point.adress,
            arrIndex: index,
          },
          {
            draggable: true,
            preset: 'islands#darkBlueDotIconWithCaptio',
            iconCaptionMaxWidth: '150',
          },
        );
        this._map.geoObjects.add(mark);
        mark.events.add(`dragend`, this.handlerDragEnd);
      });
  
      this._map.setBounds(this._map.geoObjects.getBounds(), {
        checkZoomRange: true,
        zoomMargin: 5,
      });
    }
  
  }

  addPolyLine() {
    const { points } = this.props;

    if (points.length > 0) {

      const linePoints = points.map(point => point.coords);

      const polyline = new window.ymaps.Polyline(
        linePoints,
        {
          balloonContent: 'Маршрут',
        },
        {
          balloonCloseButton: false,
          strokeColor: '#000000',
          strokeWidth: 3,
          strokeOpacity: 0.5,
        },
      );

      this._map.geoObjects.add(polyline);
    }
  }
}

const mapStateProps = state => {
  return {
    points: state.points,
  };
};

const mapDispatch = dispatch => ({
  replacePoint: dispatch.points.getNewCoords,
});

export default connect(
  mapStateProps,
  mapDispatch,
)(Map);
