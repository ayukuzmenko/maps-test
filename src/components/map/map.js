import React, { Component } from 'react';
import { connect } from 'react-redux';
import './style.css';
import { pointsSelector } from '../../selectors';

class Map extends Component {
  state = {
    map: null,
  };

  initMap = () => {
    const map = new window.ymaps.Map(
      'map',
      {
        center: [55.76, 37.64], // Москва
        zoom: 10,
      },
      {
        searchControlProvider: 'yandex#search',
      },
    );

    this.map = map;
  };

  render() {
    return <div id="map" className="containerMap " />;
  }

  componentDidMount() {
    window.ymaps.ready(this.initMap);
  }

  componentDidUpdate() {
    this.addPoint();
  }

  addPoint() {
    const { points } = this.props;

    this.map.geoObjects.removeAll();

    points.forEach(point => {
      const mark = new window.ymaps.Placemark(
        point.coords,
        {
          iconCaption: point.adress,
        },
        {
          preset: 'islands#darkBlueDotIconWithCaptio',
          iconCaptionMaxWidth: '150',
        },
      );
      this.map.geoObjects.add(mark);
    });
  }
}

const mapStateProps = state => {
  return {
    points: pointsSelector(state),
  };
};
export default connect(mapStateProps)(Map);
