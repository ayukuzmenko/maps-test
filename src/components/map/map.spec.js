import React from 'react';
import { shallow } from 'enzyme';
import { Map } from './map';
import { shallowToJson } from 'enzyme-to-json';

const mockReady = jest.fn();

window.ymaps = {
  ready: mockReady,
};

const mapContainer = shallow(<Map />);

describe('Map component test', () => {
  it('Should render map', () => {
    expect(mapContainer.find('.containerMap').first()).toBeTruthy();
    expect(mockReady).toHaveBeenCalledTimes(1);
  });

  it('Add point to map', () => {
    const newPoint = {
      adress: 'Россия, Москва',
      coords: [55.753215, 37.622504],
      id: '1554114179293.4595',
    };

    const addPoint = jest.fn();
    const addPolyLine = jest.fn();
    mapContainer.instance().addPoint = addPoint;
    mapContainer.instance().addPolyLine = addPolyLine;
    mapContainer.instance()._map = {
      geoObjects: {
        removeAll: () => {},
      },
    };
    mapContainer.update();

    mapContainer.setProps(newPoint);
    expect(addPoint).toHaveBeenCalledTimes(1);
    expect(addPolyLine).toHaveBeenCalledTimes(1);
  });

  it('Should match to snapshot', () => {
    expect(shallowToJson(mapContainer)).toMatchSnapshot();
  });
});
