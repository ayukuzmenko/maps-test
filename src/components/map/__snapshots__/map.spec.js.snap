// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map component test Should match to snapshot 1`] = `
<div
  className="containerMap"
  id="map"
/>
`;
